File: RadarSDK/RadarEfficientTrackManager.m
//
//  RadarEfficientTrackManager.m
//  RadarSDK
//
//  Created by Alan Charles on 1/21/26.
//  Copyright © 2026 Radar Labs, Inc. All rights reserved.
//

#import "RadarEfficientTrackManager.h"
#import "RadarState.h"
#import "RadarLogger.h"
#import "RadarCircleGeometry+Internal.h"
#import "RadarPolygonGeometry+Internal.h"
#import "RadarPlace+Internal.h"
#import "Radar.h"

@implementation RadarEfficientTrackManager

static const double kPlaceDetectionRadius = 100.0;

+ (BOOL)shouldTrackLocation:(CLLocation *)location options:(RadarTrackingOptions *)options {
    CLCircularRegion *syncedRegion = [RadarState syncedRegion];
    if (syncedRegion == nil) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"EfficientTrack: No synced region, should track"];
        
        return YES;
    }
    
    if ([self isOutsideSyncedRegion:location]) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"EfficientTrack: Outside synced region, should track"];
        
        return YES;
    }
    
    if ([Radar getTripOptions] != nil) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"EfficientTrack: On active trip, should always track"];
        
        return YES;
    }
    
    if (options.syncOnGeofenceEvents && [self hasGeofenceStateChanged:location]) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"EfficientTrack: Geofence state changed, should track"];
        
        return YES;
    }
    
    if (options.syncOnPlaceEvents && [self hasPlacesStateChanged:location]) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"EfficientTrack: Places state changed, should track"];
        
        return YES;
    }
    
    if (options.syncOnBeaconEvents && [self hasBeaconStateChanged:location]) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"EfficientTrack: Beacon state changed, should track"];
        
        return YES;
    }
    
    [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"EfficientTrack: No state change detected, skipping track"];
    
    return NO;
}


+ (BOOL)hasGeofenceStateChanged:(CLLocation *)location {
    NSArray<NSString *> *lastKnownGeofenceIds = [RadarState geofenceIds];
    NSArray<RadarGeofence *> *currentGeofences = [self getGeofencesForLocation: location];
    
    NSMutableSet<NSString *> *currentGeofenceIds = [NSMutableSet set];
    for (RadarGeofence *geofence in currentGeofences) {
        if (geofence._id) {
            [currentGeofenceIds addObject:geofence._id];
        }
    }
    
    NSSet<NSString *> *lastKnownSet = lastKnownGeofenceIds ? [NSSet setWithArray:lastKnownGeofenceIds] : [NSSet set];
    
    for (NSString *currentId in currentGeofenceIds) {
        if (![lastKnownSet containsObject:currentId]) {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:[NSString stringWithFormat:@"EfficientTrack: Detected geofence entry: %@", currentId]];
            
            return YES;
        }
    }
    
    for (NSString *lastKnownId in lastKnownGeofenceIds) {
        if (![currentGeofenceIds containsObject:lastKnownId]) {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:[NSString stringWithFormat:@"EfficientTrack: Detected geofence exit: %@", lastKnownId]];
            
            return YES;
        }
    }
    
    return NO;
}

+ (NSArray<RadarGeofence *> *)getGeofencesForLocation:(CLLocation *)location {
    NSArray<RadarGeofence *> *nearbyGeofences = [RadarState nearbyGeofences];
    if (nearbyGeofences == nil || nearbyGeofences.count == 0) {
        return @[];
    }
    
    NSMutableArray<RadarGeofence *> *userGeofences = [NSMutableArray array];
    
    for (RadarGeofence *geofence in nearbyGeofences) {
        RadarCoordinate *center = nil;
        double radius = 100;
        
        if ([geofence.geometry isKindOfClass:[RadarCircleGeometry class]]) {
            RadarCircleGeometry *geometry = (RadarCircleGeometry *)geofence.geometry;
            center = geometry.center;
            radius = geometry.radius;
        } else if ([geofence.geometry isKindOfClass:[RadarPolygonGeometry class]]) {
            RadarPolygonGeometry *geometry = (RadarPolygonGeometry *)geofence.geometry;
            center = geometry.center;
            radius = geometry.radius;
        } else {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:@"EfficientTrack: Skipping geofence with unsupported geometry"];
            
            continue;
        }
        
        if (center && [self isPoint:location insideCircleCenter:center.coordinate radius:radius]) {
            [userGeofences addObject:geofence];
        }
    }
    
    return  userGeofences;
}

+ (NSArray<RadarBeacon *> *)getBeaconsForLocation:(CLLocation *)location {
    NSArray<RadarBeacon *> *nearbyBeacons = [RadarState nearbyBeacons];
    if (nearbyBeacons == nil || nearbyBeacons.count == 0) {
        return @[];
    }
    
    NSMutableArray<RadarBeacon *> *userBeacons = [NSMutableArray array];
    
    static const double kBeaconRange = 100.0;
    
    for (RadarBeacon *beacon in nearbyBeacons) {
        if (beacon.geometry != nil) {
            CLLocation *beaconLocation = [[CLLocation alloc] initWithLatitude:beacon.geometry.coordinate.latitude longitude:beacon.geometry.coordinate.longitude];
            CLLocationDistance distance = [location distanceFromLocation:beaconLocation];
            
            if (distance <= kBeaconRange) {
                [userBeacons addObject:beacon];
            }
        }
    }
    
    return userBeacons;
}

+ (BOOL)hasBeaconStateChanged:(CLLocation *)location {
    NSArray<NSString *> *lastKnownBeaconIds = [RadarState beaconIds];
    NSArray<RadarBeacon *> *currentBeacons = [self getBeaconsForLocation:location];
    
    NSMutableSet<NSString *> *currentBeaconIds = [NSMutableSet set];
    for (RadarBeacon *beacon in currentBeacons) {
        if (beacon._id) {
            [currentBeaconIds addObject:beacon._id];
        }
    }
    
    NSSet<NSString *> *lastKnownSet = lastKnownBeaconIds ? [NSSet setWithArray:lastKnownBeaconIds] : [NSSet set];
    
    for (NSString *currentId in currentBeaconIds) {
        if (![lastKnownSet containsObject:currentId]) {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:[NSString stringWithFormat:@"EfficientTrack: Detected beacon entry: %@", currentId]];
            
            return YES;
        }
    }
    
    for (NSString *lastKnownId in lastKnownBeaconIds) {
        if (![currentBeaconIds containsObject:lastKnownId]) {
            [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:[NSString stringWithFormat:@"EfficientTrack: Detected beacon exit: %@", lastKnownId]];
            
            return YES;
        }
    }
    
    return NO;
}

+ (NSArray<RadarPlace *> *)getPlacesForLocation:(CLLocation *)location {
    NSArray<RadarPlace *> *nearbyPlaces = [RadarState nearbyPlaces];
    if (nearbyPlaces == nil || nearbyPlaces.count == 0) {
        return @[];
    }
    
    NSMutableArray<RadarPlace *> *userPlaces = [NSMutableArray array];
    
    for (RadarPlace *place in nearbyPlaces) {
        if (place.location != nil) {
            if ([self isPoint:location insideCircleCenter:place.location.coordinate radius:kPlaceDetectionRadius]) {
                [userPlaces addObject:place];
            }
        }
    }
    
    return userPlaces;
}

+ (BOOL)hasPlacesStateChanged:(CLLocation *)location {
    NSString *lastKnownPlaceId = [RadarState placeId];
    NSArray<RadarPlace *> *currentPlaces = [self getPlacesForLocation:location];
    
    NSMutableSet<NSString *> *currentPlaceIds = [NSMutableSet set];
    for (RadarPlace *place in currentPlaces) {
        if (place._id) {
            [currentPlaceIds addObject:place._id];
        }
    }
    
    BOOL wasInPlace = (lastKnownPlaceId != nil && lastKnownPlaceId.length > 0);
    BOOL isInPlace = [currentPlaceIds containsObject:lastKnownPlaceId];
    
    // Entry: now in a place that isn't the last known one
    if (currentPlaceIds.count > 0 && !isInPlace) {
        return YES;
    }
    
    // Exit: was in a place but no longer in any place (or different place)
    if (wasInPlace && ![currentPlaceIds containsObject:lastKnownPlaceId]) {
        return YES;
    }
    
    return NO;
}

+ (BOOL)isOutsideSyncedRegion:(CLLocation *)location {
    CLCircularRegion *syncedRegion = [RadarState syncedRegion];
    if (syncedRegion == nil) {
        return YES;
    }
    return ![syncedRegion containsCoordinate:location.coordinate];
}

+ (BOOL)isPoint:(CLLocation *)point insideCircleCenter:(CLLocationCoordinate2D)center radius:(double)radius {
    CLLocation *centerLocation = [[CLLocation alloc] initWithLatitude:center.latitude longitude:center.longitude];
    CLLocationDistance distance = [centerLocation distanceFromLocation:point];
    return distance <= radius;
}

@end

File: RadarSDK/RadarEfficientTrackManager.swift
//
//  RadarEfficientTrackManager.swift
//  RadarSDK
//
//  Created by Alan Charles on 1/29/26.
//  Copyright © 2026 Radar Labs, Inc. All rights reserved.
//

import Foundation
import CoreLocation

@objc(RadarEfficientTrackManager)
public final class RadarEfficientTrackManager: NSObject {
    
    private static let placeDetectionRadius: Double = 100.0
    private static let beaconRange: Double = 100.0
    
    @objc public static func shouldTrack(location: CLLocation, options: RadarTrackingOptions) -> Bool {
        guard RadarSwift.bridge?.syncedRegion() != nil else {
            RadarLogger.shared.debug("EfficientTrack: No synced region, should track")
            return true
        }
        
        if isOutsideSyncedRegion(location: location) {
            RadarLogger.shared.debug("EfficientTrack: Outside synced region, should track")
            return true
        }
        
        if Radar.getTripOptions() != nil {
            RadarLogger.shared.debug("EfficientTrack: On active trip, should always track")
            return true
        }
        
        if options.syncOnGeofenceEvents && hasGeofenceStateChanged(location: location) {
            RadarLogger.shared.debug("EfficientTrack: Geofence state changed, should track")
            return true
        }
        
        if options.syncOnPlaceEvents && hasPlaceStateChanged(location: location) {
            RadarLogger.shared.debug("EfficientTrack: Places state changed, should track")
            return true
        }
        
        if options.syncOnBeaconEvents && hasBeaconStateChanged(location: location) {
            RadarLogger.shared.debug("EfficientTrack: Beacon state changed, should track")
            return true
        }
        
        RadarLogger.shared.debug("EfficientTrack: No state change detected, skipping track")
        return false
    }
    
    @objc public static func getGeofences(for location: CLLocation) -> [RadarGeofence] {
        guard let nearbyGeofences = RadarSwift.bridge?.nearbyGeofences(), !nearbyGeofences.isEmpty else {
            return []
        }
        
        var userGeofences: [RadarGeofence] = []
        
        for geofence in nearbyGeofences {
            var center: RadarCoordinate?
            var radius: Double = 100
            
            if let circleGeometry = geofence.geometry as? RadarCircleGeometry {
                center = circleGeometry.center
                radius = circleGeometry.radius
            } else if let polygonGeometry = geofence.geometry as? RadarPolygonGeometry {
                center = polygonGeometry.center
                radius = polygonGeometry.radius
            } else {
                RadarLogger.shared.debug("EfficientTrack: Skipping geofence with unsupported geometry")
                continue
            }
            
            if let center = center,
               isPoint(location, insideCircleWithCenter: center.coordinate, radius: radius) {
                userGeofences.append(geofence)
            }
        }
        
        return userGeofences
    }
    
    @objc public static func getBeacons(for location: CLLocation) -> [RadarBeacon] {
        guard let nearbyBeacons = RadarSwift.bridge?.nearbyBeacons(), !nearbyBeacons.isEmpty else {
            return []
        }
        
        var userBeacons: [RadarBeacon] = []
        
        for beacon in nearbyBeacons {
            if let geometry = beacon.geometry {
                let beaconLocation = CLLocation(
                    latitude: geometry.coordinate.latitude,
                    longitude: geometry.coordinate.longitude
                )
                
                let distance = location.distance(from: beaconLocation)
                
                if distance <= beaconRange {
                    userBeacons.append(beacon)
                }
            }
        }
        
        return userBeacons
    }
    
    @objc public static func getPlaces(for location: CLLocation) -> [RadarPlace] {
        guard let nearbyPlaces = RadarSwift.bridge?.nearbyPlaces(), !nearbyPlaces.isEmpty else {
            return []
        }
        
        var userPlaces: [RadarPlace] = []
        
        for place in nearbyPlaces {
            if isPoint(location, insideCircleWithCenter: place.location.coordinate, radius: placeDetectionRadius) {
                userPlaces.append(place)
            }
        }
        
        return userPlaces
    }
    
    @objc public static func hasGeofenceStateChanged(location: CLLocation) -> Bool {
        let lastKnownGeofenceIds = RadarSwift.bridge?.geofenceIds() ?? []
        let currentGeofences = getGeofences(for: location)
        
        let currentGeofenceIds = Set(currentGeofences.compactMap { $0._id })
        let lastKnownSet = Set(lastKnownGeofenceIds)
        
        // entries
        for currentId in currentGeofenceIds {
            if !lastKnownSet.contains(currentId) {
                RadarLogger.shared.debug("EfficientTrack: Detected geofence entry: \(currentId)")
                return true
            }
        }
        
        // exits
        for lastKnownId in lastKnownGeofenceIds {
            if !currentGeofenceIds.contains(lastKnownId) {
                RadarLogger.shared.debug("EfficientTrack: Detected geofence exit: \(lastKnownId)")
                return true
            }
        }
        
        return false
    }
    
    @objc public static func hasBeaconStateChanged(location: CLLocation) -> Bool {
        let lastKNownBeaconIds = RadarSwift.bridge?.beaconIds() ?? []
        let currentBeacons = getBeacons(for: location)
        
        let currentBeaconIds = Set(currentBeacons.compactMap { $0._id })
        let lastKnownSet = Set(lastKNownBeaconIds)
        
        for currentId in currentBeaconIds {
            if !lastKnownSet.contains(currentId) {
                RadarLogger.shared.debug("EfficientTrack: Detected beacon entry: \(currentId)")
                return true
            }
        }
        
        for lastKnownId in lastKNownBeaconIds {
            if !currentBeaconIds.contains(lastKnownId) {
                RadarLogger.shared.debug("EfficientTrack: Detected beacon exit: \(lastKnownId)")
                return true
            }
        }
        
        return false
    }
    
    @objc public static func hasPlaceStateChanged(location: CLLocation) -> Bool {
        let lastKnownPlaceId = RadarSwift.bridge?.placeId()
        let currentPlaces = getPlaces(for: location)
        
        let currentPlaceIds = Set(currentPlaces.compactMap { $0._id})
        
        let wasInPlace = lastKnownPlaceId != nil && !lastKnownPlaceId!.isEmpty
        let isInPlace =  lastKnownPlaceId != nil && currentPlaceIds.contains(lastKnownPlaceId!)
        
        if !currentPlaceIds.isEmpty && !isInPlace {
            return true
        }
        
        if wasInPlace && !currentPlaceIds.contains(lastKnownPlaceId!) {
            return true
        }
        
        return false
        
    }
    
    @objc public static func isOutsideSyncedRegion(location: CLLocation) -> Bool {
        guard let syncedRegion = RadarSwift.bridge?.syncedRegion() else {
            return true
        }
        
        return !syncedRegion.contains(location.coordinate)
    }
    
    @objc public static func isPoint(_ point: CLLocation, insideCircleWithCenter center: CLLocationCoordinate2D, radius: Double) -> Bool {
        let centerLocation = CLLocation(latitude: center.latitude, longitude: center.longitude)
        let distance = centerLocation.distance(from: point)
        return distance <= radius
    }
}
